test/suites/integration/nodeclaim_test.go

```diff
@@ -0,0 +1,306 @@
+			Expect(node.Annotations).To(HaveKeyWithValue("custom-annotation", "custom-value"))
+			Expect(node.Labels).To(HaveKeyWithValue("custom-label", "custom-value"))
+			Expect(node.Spec.Taints).To(ContainElements(
+				corev1.Taint{
+					Key:    "custom-taint",
+					Effect: corev1.TaintEffectNoSchedule,
+					Value:  "custom-value",
+				},
+				corev1.Taint{
+					Key:    "other-custom-taint",
+					Effect: corev1.TaintEffectNoExecute,
+					Value:  "other-custom-value",
+				},
+			))
+			Expect(node.OwnerReferences).To(ContainElement(
+				metav1.OwnerReference{
+					APIVersion:         object.GVK(nodeClaim).GroupVersion().String(),
+					Kind:               "NodeClaim",
+					Name:               nodeClaim.Name,
+					UID:                nodeClaim.UID,
+					BlockOwnerDeletion: lo.ToPtr(true),
+				},
+			))
+			env.EventuallyExpectCreatedNodeClaimCount("==", 1)
+			env.EventuallyExpectNodeClaimsReady(nodeClaim)
+		})
+		It("should remove the cloudProvider NodeClaim when the cluster NodeClaim is deleted", func() {
+			nodeClaim := test.NodeClaim(v1.NodeClaim{
+				Spec: v1.NodeClaimSpec{
+					Requirements: requirements,
+					NodeClassRef: &v1.NodeClassReference{
+						Group: object.GVK(nodeClass).Group,
+						Kind:  object.GVK(nodeClass).Kind,
+						Name:  nodeClass.GetName(),
+					},
+				},
+			})
+			env.ExpectCreated(nodeClass, nodeClaim)
+			node := env.EventuallyExpectInitializedNodeCount("==", 1)[0]
+			nodeClaim = env.EventuallyExpectCreatedNodeClaimCount("==", 1)[0]
+
+			// Node is deleted and now should be not found
+			env.ExpectDeleted(nodeClaim)
+			env.EventuallyExpectNotFound(nodeClaim, node)
+		})
+		It("should delete a NodeClaim from the node termination finalizer", func() {
+			nodeClaim := test.NodeClaim(v1.NodeClaim{
+				Spec: v1.NodeClaimSpec{
+					Requirements: requirements,
+					NodeClassRef: &v1.NodeClassReference{
+						Group: object.GVK(nodeClass).Group,
+						Kind:  object.GVK(nodeClass).Kind,
+						Name:  nodeClass.GetName(),
+					},
+				},
+			})
+			env.ExpectCreated(nodeClass, nodeClaim)
+			node := env.EventuallyExpectInitializedNodeCount("==", 1)[0]
+			nodeClaim = env.EventuallyExpectCreatedNodeClaimCount("==", 1)[0]
+
+			// Delete the node and expect both the node and nodeClaim to be gone as well as the instance to be shutting-down
+			env.ExpectDeleted(node)
+			env.EventuallyExpectNotFound(nodeClaim, node)
+		})
+		It("should delete a NodeClaim after the registration timeout when the node doesn't register", func() {
+			nodeClaim := test.NodeClaim(v1.NodeClaim{
+				Spec: v1.NodeClaimSpec{
+					Requirements: requirements,
+					NodeClassRef: &v1.NodeClassReference{
+						Group: object.GVK(nodeClass).Group,
+						Kind:  object.GVK(nodeClass).Kind,
+						Name:  nodeClass.GetName(),
+					},
+				},
+			})
+
+			env.ExpectCreated(nodeClass, nodeClaim)
+			nodeClaim = env.EventuallyExpectCreatedNodeClaimCount("==", 1)[0]
+			nodeClaim.Status = v1.NodeClaimStatus{
```

Comment:
I will be dropping this for a validating admissions policy approach that will block any nodes from being created if there is a specific label present